src/V3MergeCond.cpp

```diff
@@ -0,0 +1,236 @@
+            }
+        }
+        rhsp->v3fatal("Don't know how to fold expression");
+    }
+
+    void mergeEnd() {
+        UASSERT(m_mgFirstp, "mergeEnd without list");
+        // Merge if list is longer than one node
+        if (m_mgFirstp != m_mgLastp) {
+            UINFO(6, "MergeCond - First: " << m_mgFirstp << " Last: " << m_mgLastp << endl);
+            ++m_statMerges;
+
+            // Create equivalent 'if' statement and insert it before the first node
+            AstIf* const ifp
+                = new AstIf(m_mgCondp->fileline(), m_mgCondp->unlinkFrBack(), NULL, NULL);
+            m_mgFirstp->replaceWith(ifp);
+            ifp->addNextHere(m_mgFirstp);
+            // Unzip the list and insert under branches
+            AstNode* nextp = m_mgFirstp;
+            do {
+                // Grab next pointer and unlink
+                AstNode* const currp = nextp;
+                nextp = currp != m_mgLastp ? currp->nextp() : NULL;
+                currp->unlinkFrBack();
+                // Skip over comments
+                if (VN_IS(currp, Comment)) {
+                    currp->deleteTree();
+                    continue;
+                }
+                // Count
+                ++m_statMergedItems;
+                // Unlink RHS and clone to get the 2 assignments (reusing currp)
+                AstNodeAssign* const thenp = VN_CAST(currp, NodeAssign);
+                AstNode* const rhsp = thenp->rhsp()->unlinkFrBack();
+                AstNodeAssign* const elsep = thenp->cloneTree(false);
+                // Construct the new RHSs and add to branches
+                thenp->rhsp(foldAndUnlink(rhsp, true));
+                elsep->rhsp(foldAndUnlink(rhsp, false));
+                ifp->addIfsp(thenp);
+                ifp->addElsesp(elsep);
+                // Cleanup
+                rhsp->deleteTree();
+            } while (nextp);
+        }
+        // Reset state
+        m_mgFirstp = NULL;
+        m_mgCondp = NULL;
+        m_mgLastp = NULL;
+        m_mgNextp = NULL;
+    }
+
+    void addToList(AstNode* nodep) {
+        // Track end of list
+        m_mgLastp = nodep;
+        m_mgNextp = nodep->nextp();
+        // Skip over comments (inserted by V3Order before always blocks)
+        while (m_mgNextp && VN_IS(m_mgNextp, Comment)) { m_mgNextp = m_mgNextp->nextp(); }
+        // If last under parent, done with current list
+        if (!m_mgNextp) { mergeEnd(); }
+    }
+
+    // VISITORS
+    virtual void visit(AstNodeAssign* nodep) VL_OVERRIDE {
+        AstNode*const rhsp = nodep->rhsp();
+        if (AstNodeCond* const condp = extractCond(rhsp)) {
+            if (m_mgFirstp && (m_mgNextp != nodep || !condp->condp()->sameTree(m_mgCondp))) {
+                // Node in different list, or has different condition
+                mergeEnd();
+            }
+            // Set up head of new list if node is first in list
+            if (m_mgFirstp == NULL) {
```

Comment:
```suggestion
            if (m_mgFirstp) {
```

However consider if cleaner to have addToList internals do all of the tracking.